File: PDOO/QytetecJava/src/modeloqytetec/Jugador.java

package modeloqytetec;

import java.util.ArrayList;

/**
 * @author Jorge Valenzuela Garcia
 */
public class Jugador {
    
    // Atributos
    private boolean encarcelado;
    String nombre;
    private int saldo;
    
    // Atributos referenciados
    private Casilla casillaActual;
    private Sorpresa cartaLibertad;
    private ArrayList<TituloPropiedad> propiedades;
    
    //Atributos especulador
    protected static int factorEspeculador = 1;
    
    // Constructor
    public Jugador(String nombre){
        this.encarcelado = false; // Cuando se empieza, nunca se está encarcelado
        this.nombre = nombre;
        this.saldo = 7500;

        this.casillaActual = new OtraCasilla(0, TipoCasilla.SALIDA);
        this.propiedades = new ArrayList();
    }
    
    // Constructor de copia
    protected Jugador(Jugador jugador){
        this.encarcelado = jugador.encarcelado;
        this.nombre = jugador.nombre;
        this.saldo = jugador.saldo;
        this.casillaActual = jugador.casillaActual;
        this.cartaLibertad = jugador.cartaLibertad;
        this.propiedades = jugador.propiedades;
    }
    
    public Casilla getCasillaActual(){
        return this.casillaActual;
    }
    
    public boolean getEncarcelado(){
        return this.encarcelado;
    }
    
    public boolean tengoPropiedades(){
        if(this.propiedades.isEmpty())
            return false;
        else
            return true;
    }
    
    protected boolean actualizarPosicion(Casilla casilla){

        if(casilla.getNumeroCasilla() < this.casillaActual.getNumeroCasilla())
            this.modificarSaldo(Qytetet.SALDO_SALIDA);
        
        boolean tienePropietario = false;
        this.setCasillaActual(casilla);
        
        if(casilla.getTipo().equals(TipoCasilla.CALLE)){
            Calle calle = (Calle) casilla;
            if(calle.soyEdificable()){
                if(calle.tengoPropietario()){
                    tienePropietario = true;
                    if(!calle.propietarioEncarcelado()){
                        this.modificarSaldo(-calle.cobrarAlquiler());
                    }
                }
            }
        }
        if(casilla.getTipo().equals(TipoCasilla.IMPUESTO)){
            this.pagarImpuestos(-casilla.getCoste());
        }
        return tienePropietario;
    }
    
    protected boolean comprarTitulo(){
        boolean puedoComprar = false;
        Calle calle = (Calle) this.casillaActual;
        if(calle.soyEdificable()){
            if(!calle.tengoPropietario()){
                if(calle.getCoste() <= this.saldo){
                    TituloPropiedad titulo = calle.asignarPropietario(this);
                    this.propiedades.add(titulo);
                    this.modificarSaldo(-calle.getCoste());
                    puedoComprar = true;
                }
            }
        }
        return puedoComprar;
    }
    
    
    protected Especulador convertirme(int fianza){
        return new Especulador(this, fianza);
    }
    
    
    protected Sorpresa devolverCartaLibertad(){ // Posible error
        Sorpresa devolver = this.cartaLibertad;
        this.cartaLibertad = null;
        return devolver;
    }
    
    protected int getFactor(){
        return Jugador.factorEspeculador;
    }
    
    public int getSaldo(){
        return this.saldo;
    }
    
    protected void irACarcel(Casilla casilla){
        this.setCasillaActual(casilla);
        this.setEncarcelado(true);
    }
    
    protected void modificarSaldo(int cantidad){
        this.saldo += cantidad;
    }
    
    protected int obtenerCapital(){
        int capital = this.saldo;
        int coste,precio,casas,hoteles;
        Calle calle;
        for(TituloPropiedad titulo : this.propiedades){
            calle = (Calle) titulo.getCasilla();
            coste = calle.getCoste();
            casas = calle.getNumCasas();
            hoteles = calle.getNumHoteles();
            precio = calle.getPrecioEdificar();
            capital += coste + ((casas+hoteles) * precio);
            if(titulo.getHipotecada())
                capital -= titulo.getHipotecaBase();
        }
        return capital;
    }
    
    protected ArrayList<TituloPropiedad> obtenerPropiedadesHipotecadas(boolean hipotecada){
        ArrayList<TituloPropiedad> devolverSi = new ArrayList();
        ArrayList<TituloPropiedad> devolverNo = new ArrayList();
        for(TituloPropiedad titulo : this.propiedades)    
            if(titulo.getHipotecada())
                devolverSi.add(titulo);
            else
                devolverNo.add(titulo);
        
        if(hipotecada)
            return devolverSi;
        else
            return devolverNo;
    }
    
    protected void pagarCobrarPorCasaYHotel(int cantidad){
        int numeroTotal = this.cuantasCasasHotelesTengo();
        this.modificarSaldo(cantidad * numeroTotal);
    }
    
    protected void pagarImpuestos(int cantidad){
        this.modificarSaldo(cantidad);
    }
    
    protected boolean pagarLibertad(int cantidad){
        boolean tengoSaldo = this.tengoSaldo(cantidad);
        if(tengoSaldo)
            this.modificarSaldo(-cantidad);
        return tengoSaldo;
    }
    
    protected boolean puedoEdificarCasa(Casilla casilla){
        // true si es de mi propiedad y tengo saldo
        if(this.esDeMiPropiedad(casilla)){
            Calle calle = (Calle) casilla;
            return this.tengoSaldo(calle.getPrecioEdificar());
        }
        return false;
    }
    
    protected boolean puedoEdificarHotel(Casilla casilla){
        // true si es de mi propiedad y tengo saldo
        if(this.esDeMiPropiedad(casilla)){
            Calle calle = (Calle) casilla;
            return this.tengoSaldo(calle.getPrecioEdificar());
        }
        else
            return false;
    }
    
    protected boolean puedoHipotecar(Casilla casilla){
        if(this.esDeMiPropiedad(casilla))
            return true;
        else
            return false;
    }
    
    protected boolean puedoPagarHipoteca(Casilla casilla){
        Calle calle = (Calle) casilla;
        if(this.esDeMiPropiedad(casilla) && this.saldo >= calle.calcularValorHipoteca())
            return true;
        else
            return false;
    }
    
    protected boolean puedoVenderPropiedad(Casilla casilla){
        boolean devolver = false;
        if(this.esDeMiPropiedad(casilla)){
            Calle calle = (Calle) casilla;
            if(!calle.estaHipotecada())
                devolver = true;
        }
        return devolver;
    }
    
    protected void setCartaLibertad(Sorpresa carta){
        if(carta.getTipo() == TipoSorpresa.SALIRCARCEL)
            this.cartaLibertad = carta;
    }
    
    protected void setCasillaActual(Casilla casilla){
        this.casillaActual = casilla;
    }
    
    protected void setEncarcelado(boolean encarcelado){
        this.encarcelado = encarcelado;
    }
    
    protected boolean tengoCartaLibertad(){
        if(this.cartaLibertad == null)
            return false;
        else
            return true;
    }
    
    protected void venderPropiedad(Casilla casilla){
        Calle calle = (Calle) casilla;
        int precioVenta = calle.venderTitulo();
        this.modificarSaldo(precioVenta);
        this.eliminarDeMisPropiedades(casilla);
    }
    
    private int cuantasCasasHotelesTengo(){
        int total = 0;
        Calle calle;
        for(TituloPropiedad titulo: this.propiedades){
            calle = (Calle) titulo.getCasilla();
            total += calle.getNumCasas();
            total += calle.getNumHoteles();
        }
        return total;
    }
    
    private void eliminarDeMisPropiedades(Casilla casilla){
        Calle calle = (Calle) casilla;
        TituloPropiedad titulo = calle.getTitulo();
        int indice;
        // Comprobamos que tenemos el titulo entre nuestras propiedades
        if(this.propiedades.contains(titulo)){
            // Obtenemos el indice de dicho titulo y borramos
            indice = this.propiedades.indexOf(titulo);
            this.propiedades.remove(indice);                    
        }
    }
    
    private boolean esDeMiPropiedad(Casilla casilla){
        boolean mio = false;
        Calle calle = (Calle) casilla;
        for(TituloPropiedad titulo : this.propiedades)
            if(calle.getTitulo().getNombre().equals(titulo.getNombre()))
                mio = true;
        return mio;
    }
            
    protected boolean tengoSaldo(int cantidad){
        return this.saldo >= cantidad;
    }
    
    
    @Override
    public String toString(){
        String salida;
        
        // Si no tengo propiedades
        if(this.propiedades.isEmpty()){
            salida =    "Nombre: " + this.nombre + "\n" +
                        "Encarcelado: " + Boolean.toString(this.encarcelado) + "\n" +
                        "Saldo: " + Integer.toString(this.saldo) + "\n" +
                        "Casilla Actual: " + this.casillaActual.toStringNumeroCasilla() + "\n";
            
            if(this.tengoCartaLibertad())
                salida += "Carta Libertad = Si\n" + "SIN PROPIEDADES";
            else
                salida += "Carta Libertad = No\n" + "SIN PROPIEDADES";
            
            return salida;
        } 
        
        // Si tengo propiedades
        else{
            salida =    "Nombre: " + this.nombre + "\n" +
                        "Encarcelado: " + Boolean.toString(this.encarcelado) + "\n" +
                        "Saldo: " + Integer.toString(this.saldo) + "\n" +
                        "Casilla Actual: " + this.casillaActual.toStringNumeroCasilla()+ "\n";
            
            if(this.tengoCartaLibertad()){
                salida += "Carta Libertad = Si\n\n";
                salida += "\tPROPIEDADES:\n";
                for (TituloPropiedad propiedad : this.propiedades)
                    salida += propiedad.toString() + "\n";
            }    
            else{
                salida += "Carta Libertad = No\n\n";
                salida += "\tPROPIEDADES:\n";
                for (TituloPropiedad propiedad : this.propiedades)
                    salida += propiedad.toString() + "\n";
                }
            
            return salida;
        }
        
    }
    
    public String toStringNombre(){
        return this.nombre;
    }
}

File: PDOO/QytetetRuby/pkg/QytetetRuby-0.0.1/lib/jugador.rb

require_relative "titulo_propiedad"
require_relative "sorpresa"
require_relative "casilla"
require_relative "tipo_casilla"
require_relative "qytetet"

module ModeloQytetet
    
  class Jugador
    def initialize(nombre)
      @encarcelado = false
      @nombre = nombre
      @saldo = 7500
      @casilla_actual = Casilla.no_titulo(0, TipoCasilla::SALIDA)
      @propiedades = Array.new
      @carta_libertad
    end
    
    attr_accessor :casilla_actual, :encarcelado
    attr_writer :carta_libertad
    
    def to_s
      puts "Nombre: #{@nombre} \n Encarcelado: #{@encarcelado} \n Saldo: #{@saldo}"
      puts "|| Casilla Actual:"
      puts @casilla_actual.to_s
      if @propiedades.empty?
        puts "Propiedades: SIN PROPIEDADES"
      else
        puts "|| Propiedades:"
        @propiedades.each do |tit|
          puts tit.to_s
        end
        puts "" #Tengo que poner esto para que el bucle funcione correctamente????
      end
    end
    
    def to_s_nombre
      puts "Nombre: #{@nombre}"
    end
    
    
    def tengo_propiedades
      if @propiedades.empty?
        return false
      else
        return true
      end
    end
    
    
    def actualizar_posicion(casilla)
      if casilla.num_casilla < @casilla_actual.num_casilla
        modificar_saldo(1000) ########### como usar la variable de qytetet saldo salida
      end
      
      tiene_propietario = false
      @casilla_actual = casilla
      
      if casilla.soy_edificable
        if casilla.tengo_propietario
          tiene_propietario = true
          if !casilla.propietario_encarcelado
            modificar_saldo(-casilla.cobrar_alquiler)
          end
        end
        
      end
      if casilla.tipo == TipoCasilla::IMPUESTO
        modificar_saldo(-100) ### Antes casilla.coste. El coste por defecto de una casilla que no es CALLE es -1 de ahi que lo cambie
      end
      return tiene_propietario
    end
    
    
    def comprar_titulo ## Como pasar el objeto this/self
      puedo_comprar = false
      if @casilla_actual.soy_edificable
        if !@casilla_actual.tengo_propietario
          if @casilla_actual.coste <= @saldo
            titulo = @casilla_actual.asignar_propietario(self) ##### Como pasar el objeto this/self
            @propiedades << titulo
            modificar_saldo(-@casilla_actual.coste)
            puedo_comprar = true
          end
        end
      end
      
      return puedo_comprar
    end
    
    
    def devolver_carta_libertad
      devolver = @carta_libertad
      @carta_libertad = nil
      return devolver
    end
    
    
    def ir_a_carcel(casilla)
      @casilla_actual = casilla
      @encarcelado = true
    end

    
    def modificar_saldo(cantidad)
      @saldo = @saldo + cantidad
    end
    
    
    def obtener_capital()
      capital = @saldo
      @propiedades.each do |titulo|
        coste = titulo.casilla.coste
        casas = titulo.casilla.num_casas
        hoteles = titulo.casilla.num_hoteles
        precio = titulo.casilla.get_precio_edificar
        capital = coste + ((casas + hoteles) * precio )
        if titulo.hipotecada
          capital = capital - titulo.hipoteca_base
        end
      end
      
      return capital.to_int;
    end
    
    
    def obtener_propiedades_hipotecadas(hipotecada)
      
      devolver_si = Array.new
      devolver_no = Array.new
      @propiedades.each do |titulo|
        if titulo.hipotecada
          devolver_si.push(titulo)
        else
          devolver_no.push(titulo)
        end
      end
      
        if hipotecada
          return devolver_si
        else
          return devolver_no
        end
              
    end
    
    
    def pagar_cobrar_por_casa_y_hotel(cantidad)
      numero_total = cuantas_casas_hoteles_tengo
      modificar_saldo(cantidad * numero_total)
    end
    
    
    def pagar_libertad(cantidad)
      tengo_saldo = tengo_saldo(cantidad)
      if tengo_saldo
        modificar_saldo(-cantidad)
      end
      return tengo_saldo
    end
    
    
    def puedo_edificar_casa(casilla)
      if es_de_mi_propiedad(casilla)
        return tengo_saldo(casilla.get_precio_edificar)
      end
      return false
    end
    
    
    def puedo_edificar_hotel(casilla)
      if es_de_mi_propiedad(casilla)
        return tengo_saldo(casilla.get_precio_edificar)
      end
      return false
    end
    
    
    def puedo_hipotecar(casilla)
      if es_de_mi_propiedad(casilla)
        return true
      else
        return false
      end
    end
    
    
    def puedo_pagar_hipoteca(casilla)
      if es_de_mi_propiedad(casilla) && @saldo >= casilla.calcular_valor_hipoteca
        return true
      else
        return false
      end
    end
    
    
    def puedo_vender_propiedad(casilla)
      devolver = false
      if es_de_mi_propiedad(casilla)
        if !casilla.esta_hipotecada
          devolver = true
        end
      end
      return devolver
    end
    
    
    def tengo_carta_libertad()
      if @carta_libertad.nil?
        return false
      else
        return true
      end
    end
    
    
    def vender_propiedad(casilla)
      precio_venta = casilla.vender_titulo
      modificar_saldo(precio_venta)
      eliminar_de_mis_propiedades(casilla)
    end
    
    
    
    private
    def cuantas_casas_hoteles_tengo
      total = 0
      @propiedades.each do |titulo|
        total = total + titulo.casilla.num_casas
        total = total + titulo.casilla.num_hoteles
      end
      return total
    end
    
    
    def eliminar_de_mis_propiedades(casilla)  ### Posible errata
      titulo = casilla.titulo                 # Cogemos el titulo de la casilla
      existe = @propiedades.include?(titulo)  # Comprobamos si existe entre las propiedades del jugador
      if existe                               # Si existe...
        @propiedades.delete(titulo)           # Lo borramos del array del propiedades
      end
    end
    
    
    def es_de_mi_propiedad(casilla)   ### Posible errata
      mio = false
      @propiedades.each do |ti|
        if casilla.titulo.eql?(ti) ## Funciona?
          mio = true
        end
      end
      return mio
    end
    
    
    def tengo_saldo(cantidad)
      if @saldo >= cantidad
        return true
      else
        return false
      end
    end
    
  end
end
